src/coreclr/jit/forwardsub.cpp

```diff
@@ -465,47 +544,86 @@ bool Compiler::fgForwardSubStatement(Statement* stmt)
         return false;
     }
 
+    // Local and tree to substitute seem suitable.
+    // See if the next statement contains the one and only use.
+    //
+    Statement* const nextStmt = stmt->GetNextStmt();
+
+    ForwardSubVisitor fsv(this, lclNum, livenessBased);
+    assert(fgNodeThreading == NodeThreading::AllLocals);
+    // Do a quick scan through the linked locals list to see if there is a last
+    // use.
+    bool found = false;
+    for (GenTreeLclVarCommon* lcl : nextStmt->LocalsTreeList())
+    {
+        if (lcl->OperIs(GT_LCL_VAR) && (lcl->GetLclNum() == lclNum))
+        {
+            if (fsv.IsLastUse(lcl->AsLclVar()))
+            {
+                found = true;
+                break;
+            }
+        }
+
+        if (fsv.IsUse(lcl))
+        {
+            JITDUMP(" next stmt has non-last use\n");
+            return false;
+        }
+    }
+
+    if (!found)
+    {
+        JITDUMP(" no next stmt use\n");
+        return false;
+    }
+
     // Don't fwd sub overly large trees.
     // Size limit here is ad-hoc. Need to tune.
     //
     // Consider instead using the height of the fwdSubNode.
     //
     unsigned const nodeLimit = 16;
 
-    if (gtComplexityExceeds(&fwdSubNode, nodeLimit))
+    if (gtComplexityExceeds(fwdSubNode, nodeLimit))
     {
         JITDUMP(" tree to sub has more than %u nodes\n", nodeLimit);
         return false;
     }
 
-    // Local and tree to substitute seem suitable.
-    // See if the next statement contains the one and only use.
-    //
-    Statement* const nextStmt = stmt->GetNextStmt();
-
     // We often see stale flags, eg call flags after inlining.
     // Try and clean these up.
     //
     gtUpdateStmtSideEffects(nextStmt);
     gtUpdateStmtSideEffects(stmt);
 
-    // Scan for the (single) use.
+    // Scan for the (last) use.
     //
-    ForwardSubVisitor fsv(this, lclNum);
     fsv.WalkTree(nextStmt->GetRootNodePointer(), nullptr);
 
-    // LclMorph (via RCS_Early) said there was just one use.
-    // It had better have gotten this right.
-    //
-    assert(fsv.GetUseCount() <= 1);
+    if (livenessBased)
+    {
+        if (fsv.GetUseCount() > 1)
```

Comment:
This is left-over from before I had the locals list scan above. I think you are right to be confused, this should be unnecessary with the above scan. I will remove it.

Reply:
Done and changed some of the other asserts around too based on the previous forward sub. The diffs were https://gist.github.com/jakobbotsch/3fc155ca8027fd4f43d2607135b5baf2.
Looks like a regression because of two large regressions in a coreclr test. I spot checked a few of the new regressions and they were all due to worse register allocation due to the new forward substitutions (bad preferencing across calls as you noted a while back, or new saves/restores in the prolog/epilog).

Reply:
Yeah, this is the IR shape I was looking at in #75565.